### PR TITLE
Fix `ttnn.matmul` For Large Inner Dim When BFloat16 Type Is Used

### LINKED ISSUE #8666: `fp32_dest_acc_en` not working properly for `ttnn.matmul`
### Describe the bug

While training different models, we have observed a significant drop in PCC after certain number of steps of training. Upon inspection using CHISEL we have deduced that the problematic op was a matmul with really high inner dim. This posed a significant blocker for training through Forge compiler.

After investigation it was noticed that we are missing proper accumulation in dest register, as copies from L1 results were not done in FP32 but in TF32, which is basically the same precision as BF16. The problematic pass is `TTNNSetComputeKernelConfig` which blindly created ComputeKernelConfig without concerning all fields required to work completely correctly.

For understanding on what is the correct approach and how we need to rework the pass I am providing links to tt-metal repo:
- https://github.com/tenstorrent/tt-metal/blob/main/docs/L1_ACCUMULATION_FP32_ANALYSIS.md
- https://github.com/tenstorrent/tt-metal/issues/28800
- https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/op_kernel_dev/accuracy_tips/accuracy_tips.md

### Steps to reproduce the issue

Tensors are too big to fit the github issue, posting the repro script for ttnn.matmul op.

[repro_chisel_matmul_ttnn_tttrain_ckcfg.py](https://github.com/user-attachments/files/28425240/repro_chisel_matmul_ttnn_tttrain_ckcfg.py)

### Expected behavior

Matmuls should properly be lowered to use fp32 accumulation.

### Please complete the following environment information

Ran on both Blackhole and Wormhole, but the repro in the issue is from Wormhole.

### PR DESCRIPTION
**Pull request by mmilosevicTT**
### Ticket
Closes https://github.com/tenstorrent/tt-mlir/issues/8666

### Problem description
`ttnn.matmul` has a correctness issue when we do not use `fp32_dest_acc_en` and `packer_l1_acc` together for BFloat16 dtype. Issue is noticable during training, as in backward graph we experience large `lm_head.weight` with `d(logits)` matmul which has really low values. This bug ends up in diverging training loss.

### What's changed
This PR includes following changes:
- Always run `TTNNSetComputeKernelConfig` pass.
- Refactor `ttnn.matmul` and `ttnn.linear` as shape adjusting of inputs is done in both version of verify, but we do not reuse code, but copy.
- Add forced setting of `fp32_dest_acc_en` and `packer_l1_acc` to `true` when inner dim of a matmul is larger than 50000.

For any details on the decision consult the issue.

### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by svuckovicTT** [comment]
fyi @rpavlovicTT

File: lib/Dialect/TTNN/IR/TTNNOps.cpp [hunk 1]
```diff
@@ -2267,6 +2267,54 @@ void mlir::tt::ttnn::ToLayoutOp::getCanonicalizationPatterns(
...
+}
+
+std::optional<int64_t> getMatmulInnerDim(::mlir::RankedTensorType inputA,
+                                         ::mlir::RankedTensorType inputB,
+                                         bool transposeA, bool transposeB) {
+  if (inputA.getRank() < 1 || inputB.getRank() < 1) {
+    return std::nullopt;
+  }
+
+  MatmulAdjustedShapes adjustedShapes =
```
Comment (sdjordjevicTT) [inline lib/Dialect/TTNN/IR/TTNNOps.cpp:2336]: Can we move this logic for adjusting vectors into a workaround decomposition pass? Also, please open a tt-metal issue as well, this is op limitation that can be handled on their end.

File: lib/Dialect/TTNN/Transforms/TTNNSetComputeKernelConfig.cpp [hunk 2]
```diff
@@ -2,11 +2,37 @@
...
+#include "ttmlir/Dialect/TTCore/IR/TTCoreOpsTypes.h"
+#include "ttmlir/Dialect/TTNN/IR/TTNNOps.h"
 #include "ttmlir/Dialect/TTNN/IR/TTNNOpsAttrs.h"
 #include "ttmlir/Dialect/TTNN/Interfaces/TTNNTensorSpecInterface.h"
 #include "ttmlir/Dialect/TTNN/Transforms/Passes.h"
 
 namespace mlir::tt::ttnn {
+namespace {
+
+constexpr int64_t kLargeInnerDimThreshold = 50000;
```
Comment (sdjordjevicTT) [inline lib/Dialect/TTNN/Transforms/TTNNSetComputeKernelConfig.cpp:19]: Please add a comment on what this constant is and in what scenarios it is important :)

Comment (mmilosevicTT) [reply lib/Dialect/TTNN/Transforms/TTNNSetComputeKernelConfig.cpp:19]: Done.

**Review by sdjordjevicTT — APPROVED** [review]

File: lib/Dialect/TTNN/IR/TTNNOps.cpp (hunk 1 above)
Comment (mmilosevicTT) [reply lib/Dialect/TTNN/IR/TTNNOps.cpp:2336]: Synced offline, this is a refactor for code maintainability.

Opened an issue for the matmul problem in tt-metal: https://github.com/tenstorrent/tt-metal/issues/46101

Comment (vladimirjovanovicTT) [reply lib/Dialect/TTNN/IR/TTNNOps.cpp:2336]: Please add a reference to the metal issue in the code.

**Comment by rpavlovicTT** [comment]
> fyi @rpavlovicTT

Thanks for notifying, looks good to me as we are limiting impact only to very large inner dim MMs.

**Comment by kmabeeTT** [comment]
(Got to drop a comment here from claude to flag something related)

**Heads-up from the inference/serving side (tt-xla vLLM).** We see a batched-prefill correctness bug that looks like the same `fp32_dest_acc_en` accumulation family: greedy decoding of N *identical* prompts in one batch diverges — the batch splits into deterministic contiguous groups that emit different tokens — and **`fp32_dest_acc_en=True` fully eliminates it** for prefill. Standalone repro + analysis: https://github.com/tenstorrent/tt-xla/issues/5116.

Relevant to this PR: the fix is gated on `inner dim > 50000` + bf16 output, but Llama inference matmul inner dims are <= ~8192, so it won't fire for inference. Consider gating by dtype/output rather than a fixed inner-dim cutoff so inference matmuls are covered.

Also worth tracking: a matmul-only fix won't cover the fused SDPA op, which has no fp32-accumulation path at all ([#8657](https://github.com/tenstorrent/tt-mlir/issues/8657)).

File: lib/Dialect/TTNN/Transforms/TTNNSetComputeKernelConfig.cpp [hunk 3]
```diff
@@ -2,11 +2,50 @@
...
+
+static bool hasSetComputeConfigFields(DeviceComputeKernelConfigAttr config) {
+  return config.getMathFidelity().has_value() || config.getMathApproxMode() ||
+         config.getFp32DestAccEn() || config.getPackerL1Acc() ||
+         config.getDstFullSyncEn();
+}
+
+template <typename OpTy>
+void maybeApplyLargeInnerDimBf16MatmulConfig(
```
Comment (dgolubovicTT) [inline lib/Dialect/TTNN/Transforms/TTNNSetComputeKernelConfig.cpp]: Nit: I don't like this maybe in the function name. It sets config conditionally, but I would just call it: applyLargeInnerDimBf16MatmulConfig...

**Review by dgolubovicTT — APPROVED** [review]
Great!

**Comment by mmilosevicTT** [comment]
Posting for reference. Change does not seem to affect perf.

https://github.com/tenstorrent/tt-xla/actions/runs/27263112306 - n150 with fix
https://github.com/tenstorrent/tt-xla/actions/runs/27263126495 - p150 with fix
https://github.com/tenstorrent/tt-xla/actions/runs/27263145450 - n150 main
https://github.com/tenstorrent/tt-xla/actions/runs/27263158358 - p150 main
